### DISCUSSION TITLE
Troubleshooting Unresponsive Play Deployments

### DISCUSSION THREAD

**Discussion by kromenak** [playframework/playframework#12031] [Q&A]
Hello, we have a Play 2.8.19 application deployed in Docker containers on AWS Elastic Container Service.

During periods of high traffic, containers sometimes report as "unhealthy," causing them to be terminated and replaced. The health check is simply a `GET version` endpoint:
```
public Result getVersion() {
    ObjectNode result = Json.newObject();
    result.put("version", BuildInfo.version());
    return ok(result);
}
```
According to ECS, the problem is that the `GET version` endpoint times out, never returning a response, three consecutive times in a row.

I'm having trouble determining the cause of this issue. Upon reviewing the Play logs, there is no warning or error indicating why the application would become unresponsive. However, this endpoint is so simple that I don't understand where/how it could be timing out unless the entire application has become unresponsive.

A few possibilities I've considered:

- I thought this might be a memory issue, but we log our memory usage to the log at a regular interval. The terminated instances appear to have normal memory usage. I've also seen a clear log message/exception if memory runs out.
- I wondered whether this could be a threading issue. I know Play purposely doesn't use too many threads, but if it stops responding to a simple endpoint like this, would that imply that all threads are busy or somehow blocked?
- The number of containers running per machine seems to affect the stability to some degree. Could there be some machine resource bottleneck we're running into? Per-container memory seems OK, but obviously the containers also share CPU and perhaps other limits like # of open files, # of threads, etc.

Ultimately, I see the symptoms of an issue, but no clear indication as to what the issue is. As a result, I'm curious if anyone in the community with more knowledge might know what the issue could be, or if there's any steps I can take to troubleshoot or get more useful logs to pinpoint the problem.

If it's helpful, a few details about our app:

- We deploy 3 containers per r4.large EC2 instance.
- We use the "MongoDB Async Java Driver" (https://mongodb.github.io/mongo-java-driver/3.12/) for database communication. This is an older driver, which was replaced by the "Reactive Streams" driver. However, I do believe it is non-blocking, though the docs can be a bit unclear about that.

**Comment by jducoeur** [comment]
How are your *other* endpoints structured?  Are they async or blocking?  If they're not explicitly async, then they're probably blocking, which would mean that you're probably thread-starved.  (High-performance Play idiom is that all entry points should be async unless they're pretty trivial.)

NB: I know Scala usage of Play far better than I do Java usage, so there's going to be a limit to how much I can chime in.  But from what you're saying, and knowing how Play works, that's what I would look for first.

**Reply by kromenak** [reply]
Hey, thanks for responding - the majority of our endpoints are async. The version endpoint is somewhat unique because it doesn't need to do any async processing.

For example, consider an endpoint to get mail for a user. We have this controller function:
```
public CompletionStage<Result> getMail(Http.Request request) {
    return mailService.getMail(getUserId(request)).thenApply(mails -> {
        return ok(Json.toJson(mails));
    });
}
```
The service might do some processing or logic, but then reaches out to the database/repository:
```
private CompletionStage<List<Mail>> getMail(String userId) {
    return generateNewMails(userId).thenCompose(a -> {
        return mailRepository.getMail(userId).thenApply(mails -> {     
            return mails;
        });
    }
```
The repository actually reaches out to the database using the MongoDB Java Async Driver:
```
public CompletableFuture<List<Mail>> getMail(String userId) {
    List<Bson> filters = new ArrayList<>();
    filters.add(eq("userId", userId));
    return findMany(MongoDbCollection.Mail, and(filters), descending("whenCreated"));
}

protected <T> CompletableFuture<List<T>> findMany(MongoCollectionInfo collection, Bson filter, Bson sort) {
    FindIterable<Document> docs = getCollection(collection).find(filter);
    if(sort != null) {
        docs = docs.sort(sort);
    }

    CompletableFuture<List<T>> future = new CompletableFuture<>();
    docs.into(new ArrayList<>(), ((List<Document> l, Throwable t) -> {
        if(t != null) {
            future.completeExceptionally(...);
        }
        else {
            try {
                future.complete(l.stream().map(document -> {
                    return (T)documentToPojo(document, collection.getPojoClass(document));
                }).collect(Collectors.toList()));
            }
            catch(Exception ex) {
                future.completeExceptionally(...));
            }
        }
    }));
    return future;
}
```
The vast majority of our code is structured in this way, which I believe is considered non-blocking and asynchronous. One thing I was a bit unsure about was the MongoDB Java Async Driver. My understanding is that it is async and non-blocking, but I'm not 100% sure about how it works internally.

**Reply by jducoeur** [reply]
Yeah, all looks appropriate on its face.  I agree that looking into the Mongo driver and confirming that it is actually non-blocking is a good idea, but nothing is jumping out at me as odd-looking here.

Any clue what your CPU usage looks like?  If that's high, it would suggest that something computationally intensive might be starving the system; if it's low, that suggests that the system is IO-bound, which in turn suggests that it is blocking *somewhere*.  (Also might be worth looking into which sorts of thread pools you have configured, and whether that matches your usage.)
